Doc reorg: Continuous Integration (#2150)
SFDO-Tooling/CumulusCI: docs/cumulusci_flow.rst

> CumulusCI Flow

Reviewer: Overarching comment: the narrative voice is not fully consistent here. In part, we speak to "you" about your capabilities and resources - "you can" or "you should". But in part, we speak in "we do", as a description of our operations. 

I would suggest that this section may be most effective if the framing is "This is what CumulusCI Flow is.  We find this value in it. Here's how it works." That is, we front-load the impetus and benefits of the process, and then speak in descriptive terms about the process, rather than about what we actually do.

==============

> CumulusCI Flow is the branching model and process we use for development of our products at Salesforce.org.

Reviewer 2: Maybe add something to explain its overall goal, like "It's the way we arrange our work to make sure that changes don't get added to the main branch of the project until they have been adequately tested, and that changes added to main get integrated into other branches in a timely fashion."

CumulusCI Flow is based on the Github Flow model with a few changes and additions.

CumulusCI Flow is implemented in the default flows provided by CumulusCI, but the approach to working with a Github repository does not require the use of CumulusCI.

.. image:: images/salesforce-org-process.png
   :scale: 50 %
   :alt: Salesforce.org Dev/Release Process Diagram

Project Considerations
----------------------
CumulusCI Flow was designed for use with Salesforce development projects, which inject some unique considerations into finding the right branching model:

> * You cannot re-cut a Salesforce managed package release with the same version as a prior release.  As a result, Git Tags are the best representation of our releases in a repository since they are a read only reference to the exact code we put into a given release.

Reviewer: Let's call out that we're talking about first-generation managed packages.

> * Most Salesforce managed package projects use Push Upgrades to keep all users of the package on the latest release.  This eliminates the need to consider supporting multiple past versions.

Reviewer: I think this bullet is overreaching. Many ISVs don't use push upgrades at all, and even we struggle with supporting previous versions.

I like David's suggestion above for how to frame the benefits of CumulusCI flow.

* Releasing managed packages has some overhead involved including manual checks by release managers to ensure nothing gets permanently locked into the package in a release.  As a result, true continuous delivery isn't an option.  Whether you're on a team that wants to deliver quickly (e.g. a two week sprint cycle) or at team that makes several larger releases a year, CumulusCI offers functionality to help cut releases for all products with any changes.

Main Builds
-----------
The main goal of the CumulusCI Flow is to always have the main branch ready to cut into a package.
This way, we can merge a fix and cut an emergency release at any time in the development process.

To test that we can package main, we upload a beta release on every commit to main and then test that beta release in a variety of Salesforce org environments concurrently.

> This build ranges from 15 minutes to 2 hours depending on the project and a passing build is proof we can package main at any point in time.

Reviewer: I'd elide the numbers; they'll vary widely. (I have one project that takes 3 hours).

When the upload of the beta release is completed, the main branch is auto-merged_ into all open feature branches.

> New betas are published on Github as a Github Release along with automatically generated release notes created by parsing the body of all Pull Requests merged since the last production release

Reviewer: ```suggestion
New betas are published on GitHub as a GitHub Release, along with automatically generated release notes drawn from the content of the Pull Requests merged since the last production release.
```

CumulusCI and Main Builds
^^^^^^^^^^^^^^^^^^^^^^^^^
There are four main flows that facilitate main builds:

> * ``ci_master``: Deploys the main branch and all dependencies into the packaging org including incrementally deleting any metadata deleted in the commit.  The end result is a package that is ready to be uploaded from the packaging org.

Reviewer: ```suggestion
* ``ci_master``: Deploys the main branch and all dependencies into the packaging org, including incrementally deleting any metadata removed since the previous deployment.  The end result is to prepare the packaging org to upload a new version.
```

> * ``release_beta``: Uploads a beta release of the code staged in the packaging org, creates a Github Tag and Release, generates release notes, adds them to the release, and merges main into all feature branches.

Reviewer: ```suggestion
* ``release_beta``: Uploads a beta release of the code staged in the packaging org, creates a GitHub Tag and Release, generates release notes, adds them to the release, and merges ``main`` into all feature branches.
```

* ``ci_beta``: Installs the beta and all dependencies into a fresh scratch org and runs the Apex tests.

> * ``ci_beta_install``: Installs the beta and all dependencies into a fresh scratch org. This is used to prepare environments for non-Apex testing such as automated browser tests.

Reviewer: This flow doesn't exist. `install_beta`?

Reviewer 2: Just take this one out; it's not particularly related to CI

Tag Naming Convention
^^^^^^^^^^^^^^^^^^^^^
CumulusCI Flow uses two naming conventions for the tags generated by the process:

* beta/1.2-Beta_3: Beta package releases
* release/1.2:  Production package releases

> By differentiating beta and release tags, we allow tooling to query for the latest beta and the latest production release of each repository.

Reviewer: I would suggest combining this with the following section and not using the word "convention", to ensure that we make clear that the tag naming is fully user-configurable.

CumulusCI and Tag Naming Convention
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
CumulusCI's default tag prefixes can be overridden if needed for particular projects by setting the values under project -> git:

* **default_branch**: Override the default branch in the repository (default: ``main``, or the current branch during ``cci project init``)
* **prefix_beta**: Override the prefix for beta tags (default: beta/)
* **prefix_feature**: Override the prefix for feature branches (default: feature/)
* **prefix_release**: Override the prefix for release tags (default: release/)

> Feature Branches

Reviewer: I think this narrative might fit better as an introduction in the parent CumulusCI Flow section. I suggest we start there with an outline in prose of how the process works, and then descend through subsections to describe the flows and mechanics that make each stage of the process go.

----------------

> Like Github Flow, CumulusCI Flow uses a simple main/feature branch model.

Reviewer: ```suggestion
Like GitHub Flow, CumulusCI Flow uses a simple main/feature branch model.
```

The main branch is the only permanent branch in the repository.
All development work (features and bug fixes) is done in feature branches prefixed with ``feature/``.

> All commits on all feature branches are tested concurrently by our CI app, MetaCI, though any CI app could be used for the same purpose.

Reviewer: I'm not sure we should be so concrete about MetaCI here. Maybe "via continuous integration, such as MetaCI or another solution"?

Once the Pull Request is merged, the feature branch is deleted.

Feature Branch Flows
^^^^^^^^^^^^^^^^^^^^
CumulusCI facilitates working with feature branches (mainly) through two default flows:

* ``dev_org``: Used to deploy the unmanaged code and all dependencies from the feature branch into a Salesforce org to create a usable development environment.
* ``ci_feature``: Deploys the unmanaged code and all dependencies into a Salesforce org (typically a fresh scratch org) and run the Apex tests. This flow is typically run by your CI app on new commits to any feature branch.

Release Branches
----------------
Some teams deliver large releases several times a year.
For this type of release cadence, Salesforce.org uses a special type of branch referred to as a release branch.
Release branches are simply a feature branch named with a number.
These long-lived branches are created off of the ``main`` branch, serve as the target branch for all
features associated with that release and are eventually merged back to the ``main`` branch when a release occurs.

> To be able to clearly track what work is associated with a specific release, release branches adhere to the following:

Reviewer: ```suggestion
To be able to clearly track what work is associated with a specific release, release branches must fulfill these criteria:
```

> * They are the parent branches of ALL feature work associated with a release. Put another way; all feature branches use the parent-child naming convention with its target release branch.

Reviewer: ```suggestion
* They are the parent branches of *all* feature work associated with a release. That is, all feature branches associated with a release are child branches of the target release branch.
```

> * Use a strict naming format: ``feature/release_num`` where ``release_num`` is a valid integer.

Reviewer: ```suggestion
* Release branches use a strict naming format: ``feature/release_num`` where ``release_num`` is a valid integer.
```

> Using ``feature/`` branch prefix for the release branch names allow those branches to stay in sync with our main branch (they are just another feature branch to CumulusCI).

Reviewer: ```suggestion
Using the ``feature/`` branch prefix for release branch names allow those branches to stay in sync with the main branch. Like any other feature branch, they participate in CumulusCI's parent-to-child merge operations.
```

The release number immediately after the ``feature/`` prefix allows CumulusCI to detect and merge changes from one release branch to other future release branches.
See `Release to (Future) Release Merges`_ for more information.

An example of a release branch with two items of work associated with it could look like this:

* ``feature/001``
* ``feature/001__feature1``
* ``feature/001__feature2``

Branch Configuration
--------------------

> The name of the main (default) branch, as well as the branch prefixes are configurable in your projects ``cumulusci.yml`` file.

Reviewer: This section may be duplicative of the one above.

The following shows the default values that CumulusCI comes with:

.. code-block:: yaml

   project:
      git:
         default_branch: main
         prefix_feature: feature/
         prefix_beta: beta/
         prefix_release: release/

These values can be changed to match naming conventions used by your own project.

.. _auto-merged:

Auto Merging
------------

> CumulusCI helps to keep large diffs and merge conflicts from being the norm. CumulusCI's auto-merge functionality helps teams:

Reviewer: ```suggestion
CumulusCI Flow helps to keep large diffs and merge conflicts from being the norm. CumulusCI's auto-merge functionality helps teams:
```

This is also a key benefit we could articulate in the introduction, and then reiterate here.

* Keep feature branches up-to-date with the ``main`` branch (main to feature merges)
* Manage long-lived feature branches for larger features worked on by multiple developers (parent to child merges)
* Manage large releases that occur several times a year (release to future release merges).  

Main to Feature Merges 
^^^^^^^^^^^^^^^^^^^^^^

> One of the bigger differences between CumulusCI Flow and Github Flow or git-flow is that we automate the merging of commits to a projects main branch into all open feature branches.  This auto-merge does a lot for us:

Reviewer: ```suggestion
One of the bigger differences between CumulusCI Flow and GitHub Flow or git-flow is that CumulusCI Flow automates the merging of commits to a project's main branch into all open feature branches.  This auto-merge does a lot for us:
```

* Ensures feature branches are in sync with the  main branch.

> * Re-tests each feature branch with any changes to main since the merge generates a new commit

Reviewer: ```suggestion
* Re-tests each feature branch with any changes to main, since the merge generates a new commit.
```

* Eliminates merge conflicts when merging a Pull Request to main

To understand the benefit of auto-merging to feature branches, consider the following scenario: A developer starts work on a feature branch, puts in a few weeks on it, and then has to leave unexpectedly for a few months.  While they are on leave, their feature branch gets automatically updated with any new commits on main and rebuilt.  A few weeks into their leave, a new commit on main gets merged to their feature branch and breaks the build.  When the developer returns after their leave, they can look at the build history to find which commit from main broke their feature branch.

Without auto-merging, the developer would return, merge main into their feature branch, and then have to sift through all the commits to main during their leave to figure out which one broke their feature branch.  More testing and build history is always a good thing in addition to the other benefits we gain from auto-merging.

> CumulusCI facilitates the auto-merge to feature branches via the ``github_automerge_main`` task which is included by default in the ``release_beta`` flow.

Reviewer: ```suggestion
CumulusCI facilitates the auto-merge to feature branches via the ``github_automerge_main`` task, which is included by default in the ``release_beta`` flow. The ``release_beta`` flow is run, in CumulusCI Flow, on new commits to the main branch.
```

Parent to Child Merges
^^^^^^^^^^^^^^^^^^^^^^

> As we've worked in the CumulusCI Flow for the last 4+ years, we've occasionally seen the need for longer running, collaborative feature branches that are used by multiple developers to work on different parts of a single large feature. The solution was to expand the concept of auto-merging main-to-feature branches to also handle the concept of Parent and Child Feature Branches.

Reviewer: I would suggest phrasing in terms of the process and capability rather than our history.

Parent/Child Feature Branches are created using a simple naming format for branches:

* **Parent**: feature/parent-branch-name
* **Child**: feature/parent-branch-name__child-branch-name

> If this combination of named parent and child branches exist, the auto-merging functionality changes a bit:

Reviewer: Is this really a change? It seems more like just an elaboration of the automerge we've talked about.

The parent to child merge functionality works across multiple levels of branching. The effects of automerging remains the same, with children only receiving merges from their parents only (e.g. no merges from grandparents)
This allows us to have branching structures such as:

   * ``main``
   * ``feature/large-feature``
   * ``feature/large-feature__section1``
   * ``feature/large-feature__section1__work-item1``
   * ``feature/large-feature__section1__work-item2``
   * ``feature/large-feature__section2``
   * ``feature/large-feature__section2__work-item1``

In this scenario, a commit to the ``main`` branch triggers the ``github_automerge_main`` task to run and will automerge that commit into ``feature/large-feature``.
This triggers a build to run against ``feature/large-feature``, and assuming the build passes, runs the ``github_automerge_feature`` task.
This task detects two child branches of ``feature/large-feature``; ``feature/large_feature__section1`` and ``feature/large-feature__section2``.
The task automerges the commit from the parent, into the child branches, and builds begin to run against those branches.

> If the build for ``feature/large-feature__section1`` fails; then it would not trigger ``github_automerge_feature`` against it.

Reviewer: ```suggestion
If the build for ``feature/large-feature__section1`` fails,  it does not trigger ``github_automerge_feature`` to merge to its child branches.
```

> This means that despite ``feature/large-feature__section1`` having two child branches, they would not receive automerges.

Reviewer: ```suggestion
This means that despite ``feature/large-feature__section1`` having two child branches, they would not receive automerges until the parent branch tests successfully.
```

Consider the following branches in a GitHub repository:

   * ``main`` - Source of Truth for Production
   * ``feature/002`` - The next major production release
   * ``feature/002__feature1`` - A single feature associated with release ``002``
   * ``feature/002__large_feature`` - A large feature associated with release ``002``
   * ``feature/002__large_feature__child1`` - First chunk of work for the large feature
   * ``feature/002__large_feature__child2`` - Second chunk of work for the large feature
   * ``feature/003`` - The release that comes after ``002``
   * ``feature/003__feature1`` - A single feature associated with release ``003``

In this scenario, CumulusCI ensures that when ``feature/002`` receives a commit, that that commit is also merged into ``feature/003``.
This kicks off tests in our CI system and ensures that functionality going into ``feature/002`` doesn't break work being done for future releases.
Once those tests pass, the commit on ``feature/003`` is merged to ``feature/003__feature1`` because they adhere to the parent/child naming convention described above.

> Commits **never** propagate in the opposite direction. (A commit to ``feature/002`` would never be merged to ``feature/001`` if it was an existing branch in the GitHub repository).

Reviewer: Just calling out that I really like these scenarios and merge diagrams. I think this is extremely helpful for understanding a much more complex automerge landscape.

**Propagating commits to future release branches is turned off by default.** 
If you would like to enable this feature for your GitHub repository, you can set the ``update_future_releases`` option on the ``github_automerge_feature`` task in your ``cumulusci.yml`` file: 

.. code-block:: yaml 

   tasks:
      github_automerge_feature:
      options:
         update_future_releases: True

Orphan Branches
^^^^^^^^^^^^^^^
If you have both a parent and a child branch, and the parent is deleted, this creates an orphaned branch.
Orphaned branches do not receive any auto-merges from any branches.
You can rename an orphaned branch to include the ``feature/`` prefix and contain no double underscores ('__') to begin receiving merges from the main branch again.

If we have a parent and child branch: ``feature/myFeature`` and ``feature/myFeature__child``, and ``feature/myFeature`` (the parent) is deleted, then ``feature/myFeature__child`` would be considered an orphan.
Renaming ``feature/myFeature__child`` to ``feature/child`` will allow the orphan to begin receiving automerges from the main branch.

> CumulusCI Flow vs Github Flow

Reviewer: ```suggestion
CumulusCI Flow vs. GitHub Flow
```

-----------------------------

> Since CumulusCI Flow is largely an extension of Github Flow, the differences are mostly additional process in CumulusCI Flow that's not in Github Flow:

Reviewer: ```suggestion
Since CumulusCI Flow is largely an extension of GitHub Flow, the differences are mostly additional processes in CumulusCI Flow that help make it more effective for large-scale Salesforce projects:
```

* Feature branches must be prefixed feature/ or they don't get built or receive auto-merges.  This allows developers to have experimental branches that don't get built or merged.

> * CumulusCI Flow is focused on an agile release process (we use 2 week sprints/releases) instead of continuous delivery.

Reviewer: ```suggestion
* CumulusCI Flow is focused on an agile release process that works well with the technical constraints of Salesforce packaging.
```

* CumulusCI Flow requires the beta and release tag naming convention so tooling can use the Github API to determine the latest beta and the latest production release.

> * Github Flow does not do any auto-merging of commits which is a core part of CumulusCI Flow

Reviewer: I'd flip this around and articulate it as a positive of CumulusCI Flow.

> * Github Flow does not have any concept of parent/child branches though they could be manually created and maintained

Reviewer: This could fold into the previous bullet.

CumulusCI Flow vs git-flow
--------------------------

> When I first started figuring out our development/release process, I started where most people do in looking at git-flow.  Unlike both CumulusCI Flow and Github Flow, git-flow uses multiple permanent branches to separate development work from releases.  We decided to go with a main/feature branching model instead of git-flow for a few reasons:

Reviewer: Shifting personas here. I don't think we should utilize first person narration.

  As a result, true continuous delivery isn't an option.
  Whether you're on a team that wants to deliver quickly (e.g. a two week sprint cycle) or at team that makes
  several larger releases a year, CumulusCI offers functionality to help cut releases for all products with any changes.

Main Builds
-----------
The main goal of the CumulusCI Flow is to always have the main branch ready to cut into a package.
This way, we can merge a fix and cut an emergency release at any time in the development process.

To test that we can package main, we upload a beta release on every commit to main and then test that
beta release in a variety of Salesforce org environments concurrently.
Build times vary widely depending on the project and, a passing build is proof we can package main at any point in time.

When the upload of the beta release is completed, the main branch is auto-merged_ into all open feature branches.
New betas are published on GitHub as a GitHub Release, along with automatically generated release
notes drawn from the content of the Pull Requests merged since the last production release.

CumulusCI and Main Builds
^^^^^^^^^^^^^^^^^^^^^^^^^

> There are four main flows that facilitate main builds:

Reviewer: ```suggestion
There are three main flows that facilitate main builds:
```

.. _CumulusCI Flow:

CumulusCI Flow
==============
CumulusCI Flow is the process that Salesforce.org uses to: develop, test, and release, our products.
This process encompases both a development and testing philosophy as well as a specific GitHub branching structure.
There are several key reasons we like using CumulusCI Flow:

* Everything done in scratch orgs to eliminate "the org problem" <TODO link?>.  The only persistent org in this process is the packaging org. 
* Changes to branches that are being actively developed are thoroughly tested on each commit.
* A new Beta Version of the package is created and tested for each commit on a project's ``main`` branch. 
* Auto-merging functionality keeps branches up-to-date with the ``main`` branch.

> CumulusCI Flow is implemented in the out-of-the-box flows provided by CumulusCI, but the approach to working

Author: This sentence is awkward with 'flows' mentioned twice.
